src/compiler/factory/emitHelpers.ts

```diff
@@ -783,7 +783,11 @@ namespace ts {
         text: `
             var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
                 if (k2 === undefined) k2 = k;
-                Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
+                var oldDesc = Object.getOwnPropertyDescriptor(m, k);
+                var desc = (oldDesc.writable === false && oldDesc.configurable === false
+                            || m.__esModule && "get" in oldDesc)
+                           ? oldDesc : { enumerable: true, get: function() { return m[k]; } };

```

Comment:
I'm curious about the conditions here:
- `oldDesc.configurable === false` could just be `!oldDesc.configurable`, because `configurable` is always set by the runtime.
- `oldDesc.writable === false` means there won't be a `"get"` in oldDesc, because `writable` is only added for data properties (and is `undefined` for accessor properties).
- There's a possibility that `oldDesc` is undefined (if re-exporting something from a dynamic module), in which case `oldDesc.writable` could throw

Would the following condition also match the intended behavior here?

```suggestion
                var desc = oldDesc && ("get" in oldDesc ? m.__esModule : !oldDesc.writable && !oldDesc.configurable)
                           ? oldDesc : { enumerable: true, get: function() { return m[k]; } };
```

By branching on `"get" in oldDesc"` early, we can just use `!` for the `writable`/`configurable` tests, since `oldDesc.writable` will *never* be `false` (or `true`) for an accessor.

Reply:
Alternatively, this could be rewritten into an `if` and a reassignment (which is a bit more readable):
```js
var desc = Object.getOwnPropertyDescriptor(m, k);
if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
  desc = { enumerable: true, get: function() { return m[k]; } };
}
Object.defineProperty(o, k2, desc);
```

Reply:
All of that makes sense, will push that in a minute.

* Re using `!`, I guess that it makes sense since it's part of the descriptor spec (I was focused on actual node behavior wrt cjs bindings from a frozen `exports`, which I wanted to make as robust as possible).
* `writable` vs `get` -- makes sense.
* And the reassignment is a nice improvement though the resulting negation of the condition makes it more confusing, so I don't care either way.